site/app/models/gradeable/TaGradedGradeable.php

```diff
@@ -91,6 +93,116 @@ public function toArray() {
         return $details;
     }
 
+    /**
+     * Gets all component grades for a given component
+     * @param int $component_id The id of the component to get grades for
+     * @return GradedComponent[] An array of component grades (empty if non exist)
+     */
+    public function getGradedComponentsByComponentId($component_id) {
+        return $this->graded_components[$component_id] ?? [];
+    }
+
+    /**
+     * Used to retrieve existing graded components or generate new ones
+     * This function has fairly complex behavior to achieve a large amount of convenience.
+     * In general: If the component is peer or generate is true, don't pass a null grader
+     *
+     * Grader Null:
+     *   Component not peer:
+     *     Component has grades:            => return the one TA grade
+     *     Generate false:                  => return null
+     *     Generate true:                   => throw InvalidArgumentException
+     *   Component peer:                    => throw InvalidArgumentException
+     * Grader Not Null:
+     *   Component not peer:
+     *     Component has grades:            => return the one TA grade and sets the grader
+     *     Generate false:                  => return null
+     *     Generate true:                   => return new component with provided user as grader (TA)
```

Comment:
All of this creation seems like it would work better in its own function. I wouldn't expect a getter to create and set values. I would assume it returns null and I have to create the GradedComponent myself.

Reply:
The purpose of these kinds of getters is so the caller doesn't always have to be doing the "if exists, then create" check (see 1st PR comment).  In this particular case, when you're changing the grade for the assignment, there will never be a case where you want to not create the grade if it doesn't exist.

Reply:
While true, this gets rather confusing. Even in your test code, you call a getter method to create (set) a graded component and it's pretty counterintuitive. The lines of code you save probably aren't worth the confusion of a getter that is also a setter. In my opinion you should really make getters non-modifying and have createOrGet functions if the convenience is really necessary.

Reply:
I agree.  I'll give it a name change (and the other one below)